src/iceberg/snapshot.h

```diff
@@ -0,0 +1,217 @@
+
+/// \brief The Write-Audit-Publish id of a staged snapshot
+constexpr std::string_view kWAPIDKey = "wap.id";
+/// \brief The Write-Audit-Publish id of a snapshot already been published
+constexpr std::string_view kPublishedWAPIDKey = "published-wap-id";
+/// \brief The original id of a cherry-picked snapshot
+constexpr std::string_view kSourceSnapshotIDKey = "source-snapshot-id";
+/// \brief Name of the engine that created the snapshot
+constexpr std::string_view kEngineNameKey = "engine-name";
+/// \brief Version of the engine that created the snapshot
+constexpr std::string_view kEngineVersionKey = "engine-version";
+
+/// \brief The operation field is used by some operations, like snapshot expiration, to
+/// skip processing certain snapshots.
+enum class Operation {
+  /// Only data files were added and no files were removed.
+  kAppend,
+  /// Data and delete files were added and removed without changing table data; i.e.
+  /// compaction, change the data file format, or relocating data files.
+  kReplace,
+  /// Data and delete files were added and removed in a logical overwrite operation.
+  kOverwrite,
+  /// Data files were removed and their contents logically deleted and/or delete files
+  /// were added to delete rows.
+  kDelete,
+};
+
+/// \brief Summarises the changes in the snapshot.
+class ICEBERG_EXPORT Summary : public iceberg::util::Formattable {
+ public:
+  Summary() = default;
+  /// \brief Construct a summary with the given operation and properties.
+  Summary(Operation op, std::unordered_map<std::string, std::string> props);
+
+  /// \brief Get the operation type of the snapshot.
+  [[nodiscard]] Operation operation() const;
+
+  /// \brief Get the additional properties of the snapshot.
+  [[nodiscard]] const std::unordered_map<std::string, std::string>& properties() const;
+
+  std::string ToString() const override;
+
+  friend bool operator==(const Summary& lhs, const Summary& rhs) {
+    return lhs.Equals(rhs);
+  }
+
+  friend bool operator!=(const Summary& lhs, const Summary& rhs) { return !(lhs == rhs); }
+
+ private:
+  /// \brief Compare two Summaries for equality.
+  [[nodiscard]] bool Equals(const Summary& other) const;
+
+  /// The type of operation in the snapshot
+  Operation operation_{Operation::kAppend};
+  /// Other summary data.
+  std::unordered_map<std::string, std::string> additional_properties_;
+};
+
+/// \brief A snapshot of the data in a table at a point in time.
+///
+/// A snapshot consist of one or more file manifests, and the complete table contents is
+/// the union of all the data files in those manifests.
+///
+/// Snapshots are created by table operations.
+class ICEBERG_EXPORT Snapshot : public iceberg::util::Formattable {
+ public:
+  Snapshot(int64_t snapshot_id, std::optional<int64_t> parent_snapshot_id,
+           int64_t sequence_number, int64_t timestamp_ms, std::string manifest_list,
+           std::shared_ptr<Summary> summary, std::optional<int64_t> schema_id);
```

Comment:
Will a Summary actually get shared across multiple snapshots? Shouldn't it be unique_ptr or even better just use Summary directly?